pkg/sql/catalog/descs/hydrate.go

```diff
@@ -120,6 +121,19 @@ func makeMutableTypeLookupFunc(
 		mut.UpsertDescriptorEntry(desc)
 	}
 	mutableLookupFunc := func(ctx context.Context, id descpb.ID) (catalog.Descriptor, error) {
+		// This special case exists to deal with the desire to use enums in the
+		// system, and the fact that the hydration contract is such that when we
+		// resolve types mutably, we resolve the mutable type descriptors they
+		// reference, which may have in-memory changes. The problem with this is
+		// that one is not permitted to mutably resolve the public schema
+		// descriptor for the system database. We only want it for the name, so
+		// let the caller have the immutable copy.
+		if id == catconstants.PublicSchemaID {
+			return tc.GetImmutableDescriptorByID(ctx, txn, id, tree.CommonLookupFlags{
+				Required:    true,
+				AvoidLeased: true,
+			})
+		}
```

Comment:
I worry that, whenever we do endow system.public with an actual descriptor, removing this code will be overlooked.  I'm not sure what to do about this, though. Perhaps me referencing https://github.com/cockroachdb/cockroach/issues/89849 is enough.